File: src/maptoposter/rendering/road_styles.py
"""Road styling based on highway type hierarchy."""

from typing import List, Dict, Any


def get_edge_colors_by_type(G, theme: Dict[str, Any]) -> List[str]:
    """
    Assigns colors to edges based on road type hierarchy.

    Args:
        G: NetworkX graph from OSMnx
        theme: Theme dictionary with road color definitions

    Returns:
        List of colors corresponding to each edge in the graph
    """
    edge_colors = []

    for u, v, data in G.edges(data=True):
        highway = data.get('highway', 'unclassified')

        # Handle list of highway types (take the first one)
        if isinstance(highway, list):
            highway = highway[0] if highway else 'unclassified'

        # Assign color based on road type
        if highway in ['motorway', 'motorway_link']:
            color = theme['road_motorway']
        elif highway in ['trunk', 'trunk_link', 'primary', 'primary_link']:
            color = theme['road_primary']
        elif highway in ['secondary', 'secondary_link']:
            color = theme['road_secondary']
        elif highway in ['tertiary', 'tertiary_link']:
            color = theme['road_tertiary']
        elif highway in ['residential', 'living_street', 'unclassified']:
            color = theme['road_residential']
        else:
            color = theme['road_default']

        edge_colors.append(color)

    return edge_colors


def get_edge_widths_by_type(G) -> List[float]:
    """
    Assigns line widths to edges based on road type.
    Major roads get thicker lines.

    Args:
        G: NetworkX graph from OSMnx

    Returns:
        List of line widths corresponding to each edge
    """
    edge_widths = []

    for u, v, data in G.edges(data=True):
        highway = data.get('highway', 'unclassified')

        if isinstance(highway, list):
            highway = highway[0] if highway else 'unclassified'

        # Assign width based on road importance
        if highway in ['motorway', 'motorway_link']:
            width = 1.2
        elif highway in ['trunk', 'trunk_link', 'primary', 'primary_link']:
            width = 1.0
        elif highway in ['secondary', 'secondary_link']:
            width = 0.8
        elif highway in ['tertiary', 'tertiary_link']:
            width = 0.6
        else:
            width = 0.4

        edge_widths.append(width)

    return edge_widths

File: maptoposter-rs/src/rendering/road_styles.rs
use crate::core::osm_client::HighwayType;

/// Road style configuration
#[derive(Debug, Clone)]
pub struct RoadStyle {
    pub width: f32,
    pub color_key: &'static str,
    pub default_color: &'static str,
}

impl RoadStyle {
    pub fn for_highway(highway_type: HighwayType) -> Self {
        match highway_type {
            HighwayType::Motorway | HighwayType::MotorwayLink => Self {
                width: 1.2,
                color_key: "road_motorway",
                default_color: "#0A0A0A",
            },
            HighwayType::Trunk | HighwayType::Primary | HighwayType::PrimaryLink => Self {
                width: 1.0,
                color_key: "road_primary",
                default_color: "#1A1A1A",
            },
            HighwayType::Secondary | HighwayType::SecondaryLink => Self {
                width: 0.8,
                color_key: "road_secondary",
                default_color: "#2A2A2A",
            },
            HighwayType::Tertiary | HighwayType::TertiaryLink => Self {
                width: 0.6,
                color_key: "road_tertiary",
                default_color: "#3A3A3A",
            },
            HighwayType::Residential | HighwayType::LivingStreet => Self {
                width: 0.4,
                color_key: "road_residential",
                default_color: "#4A4A4A",
            },
            HighwayType::Service | HighwayType::Unclassified => Self {
                width: 0.3,
                color_key: "road_residential",
                default_color: "#4A4A4A",
            },
            HighwayType::Default => Self {
                width: 0.4,
                color_key: "road_default",
                default_color: "#3A3A3A",
            },
        }
    }
}

/// Get the drawing priority for a highway type (higher = drawn later = on top)
pub fn highway_priority(highway_type: HighwayType) -> u8 {
    match highway_type {
        HighwayType::Motorway | HighwayType::MotorwayLink => 10,
        HighwayType::Trunk => 9,
        HighwayType::Primary | HighwayType::PrimaryLink => 8,
        HighwayType::Secondary | HighwayType::SecondaryLink => 6,
        HighwayType::Tertiary | HighwayType::TertiaryLink => 4,
        HighwayType::Residential | HighwayType::LivingStreet => 2,
        HighwayType::Service | HighwayType::Unclassified | HighwayType::Default => 1,
    }
}
